src/compression/lz77.rs

```diff
@@ -0,0 +1,130 @@
+pub fn lz77_encode(input: &str) -> Vec<(usize, usize, char)> {
+    let mut tokens = Vec::new();
+    let mut index = 0;
+
+    while index < input.len() {
+        let mut best_match = (0, 0);
+        let candidate = &input[index..input.len() - 1];
+
+        for i in 0..index {
+            let search_box = &input[i..index];
+            let match_length = candidate
+                .chars()
+                .zip(search_box.chars())
+                .take_while(|&(a, b)| a == b)
+                .count();
+
+            if match_length > best_match.1 {
+                best_match = (index - i, match_length);
+            }
+        }
+
+        if best_match.1 > 0 {
+            tokens.push((
+                best_match.0,
+                best_match.1,
+                input.chars().nth(index + best_match.1).unwrap(),
+            ));
+            index += best_match.1 + 1;
+        } else {
+            tokens.push((0, 0, input.chars().nth(index).unwrap()));
+            index += 1;
+        }
+    }
+    tokens
+}
+
+pub fn lz77_decode(tokens: Vec<(usize, usize, char)>) -> String {
+    let mut result = String::new();
+    for token in tokens {
+        if token.0 != 0 {
+            let start = result.len() - token.0;
+            let length = token.1;
+            let substring: String = result.chars().skip(start).take(length).collect();
+            result += &substring;
+        };
+        result.push(token.2);
+    }
+    result
+}
```

Comment:
```suggestion
```suggestion
/// Decompresses a vector of tokens generated by the LZ77 algorithm and returns the original input string.
///
/// # Arguments
///
/// * `tokens` - A vector of tokens generated by the LZ77 compression algorithm.
///
/// # Returns
///
/// The original input string before compression.
pub fn lz77_decode(tokens: Vec<Token>) -> String {
    let mut result = String::new();
    for token in tokens {
        if token.distance != 0 {
            let start = result.len() - token.distance;
            let length = token.length;
            let substring: String = result.chars().skip(start).take(length).collect();
            result += &substring;
        };
        result.push(token.next_char);
    }
    result
}
```
```